File: CodePractice/双指针/有效回文.py
"""
给定一个字符串 s ，验证 s 是否是 回文串 ，只考虑字母和数字字符，可以忽略字母的大小写。

本题中，将空字符串定义为有效的 回文串 。

 

示例 1:

输入: s = "A man, a plan, a canal: Panama"
输出: true
解释："amanaplanacanalpanama" 是回文串
示例 2:

输入: s = "race a car"
输出: false
解释："raceacar" 不是回文串
 

提示：

1 <= s.length <= 2 * 105
字符串 s 由 ASCII 字符组成

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/XltzEq
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""

class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1

        while left < right:
            # check valid char
            while left < right and not self.is_valid(s[left]):
                left += 1
            while left < right and not self.is_valid(s[right]):
                right -= 1

            # check two char is same
            if left < right and s[left].lower() != s[right].lower():
                return False

            # continue to next
            left += 1
            right -= 1

        return True

    @staticmethod
    def is_valid(char: str) -> bool:
        return char.isdigit() or char.isalpha()

File: CodePractice/双指针/有效回文.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        int left = 0, right = s.length() - 1;

        while (left < right) {
            while (left < right && !my_isalnum(s[left]))
                ++left;
            while (left < right && !my_isalnum(s[right]))
                --right;
            if (left < right && tolower(s[left]) != tolower(s[right])) // or -32 if upper
                return false;
            ++left; --right;
        }

        return true;
    }

    bool my_isalnum(char ch) {
        return isalnum(static_cast<unsigned char>(ch));
    }

    bool my_isvalid(const char ch) {
        // 97-122  65-90  48-57
        if (ch >= 'A' && ch <= 'Z' ||
            ch >= 'a' && ch <= 'z' ||
            ch >= '0' && ch <= '9')
            return true;
        return false;
    }
};


int main() {
    Solution s = Solution();

    s.isPalindrome("A man, a plan, a canal: Panama");

    return 0;
}